moped-editor/src/views/projects/projectView/ToolbarPlugin.js

```diff
@@ -0,0 +1,290 @@
+      editor.dispatchCommand(REMOVE_LIST_COMMAND) :
+      editor.dispatchCommand(command);
+  };
+
+  const checkLink = (selection) => {
+    let isLink = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      if ($isLinkNode(selectedNodeParent)) {
+        isLink = true
+      }
+    })
+    return isLink;
+  };
+
+  const handleLinkUpdate = () => {
+    const isLinkApplied = selectionMap["link"];
+    isLinkApplied ?
+      editor.dispatchCommand(TOGGLE_LINK_COMMAND, null)
+      :
+      navigator.clipboard.readText().then(val => {
+        editor.dispatchCommand(TOGGLE_LINK_COMMAND, {
+          url: val, target: "_blank"
+        })
+      }).catch(err => {
+        console.error('Failed to read clipboard contents: ', err);
+      })
+  };
+
+  const clearFormatting = () => {
+    const formatTypes = Object.keys(selectionMap);
+    formatTypes.forEach((formatType) => {
+      if (selectionMap[formatType]) {
+        onAction(formatType);
+      }
+    });
+  };
+
+  const updateToolbar = useCallback(() => {
+    const selection = $getSelection();
+    if ($isRangeSelection(selection)) {
+      const newSelectionMap = {
+        [richTextAction.bold]: selection.hasFormat("bold"),
+        [richTextAction.italics]: selection.hasFormat("italic"),
+        [richTextAction.underline]: selection.hasFormat("underline"),
+        [richTextAction.number]: checkListType(selection, "number"),
+        [richTextAction.bullet]: checkListType(selection, "bullet"),
+        [richTextAction.link]: checkLink(selection),
+      }
+      setSelectionMap(newSelectionMap)
+    }
+  }, [setSelectionMap])
+
+  const getSelectedButtonProps = (isSelected) =>
+    isSelected
+      ? {
+        className: classes.toolbarButtons
+      }
+      : {};
+
+  const onAction = (id) => {
+    switch (id) {
+      case "bold":
+        editor.dispatchCommand(FORMAT_TEXT_COMMAND, "bold");
+        break;
+      case "italics":
+        editor.dispatchCommand(FORMAT_TEXT_COMMAND, "italic");
+        break;
+      case "underline":
+        editor.dispatchCommand(FORMAT_TEXT_COMMAND, "underline");
+        break;
+      case "number":
+        handleListUpdate(INSERT_ORDERED_LIST_COMMAND, "number");
+        break;
+      case "bullet":
+        handleListUpdate(INSERT_UNORDERED_LIST_COMMAND, "bullet");
+        break;
+      case "link":
+        handleLinkUpdate();
+        break;
+      case "undo":
+        editor.dispatchCommand(UNDO_COMMAND, undefined);
+        break;
+      case "redo":
+        editor.dispatchCommand(REDO_COMMAND, undefined);
+        break;
+      case "clear":
+        clearFormatting();
+        break;
+      default:
+        break;
+    }
+  }
+
+  // Clear editor formatting when note is saved successfully. Prop is passed down from
+  // parent component so the toolbar will rerender on save
+  if (noteAddSuccess) {
```

Comment:
perhaps it's a bit awkward to have this logic free floating here before the `useEffect`, but it functions just as it would in the `useEffect` since the prop is passed down from the parent component. placing this inside the `useEffect` causes a continuous loop which can be fixed but may require a refactor of the parent component. not sure if this is worth tackling in this pr but wanted to point that out and welcome any thoughts!